Introduces CGWindowList API for window capturing and improves `CGImage` and `CGContext`

application-developer-DA opened the pull request:
Initially I just wanted to rewrite https://github.com/servo/core-graphics-rs/pull/52 as it seems that the author is not interested in changing his pull request. So I just implemented `CGImage` bindings, but today another pull request has been merged (https://github.com/servo/core-graphics-rs/pull/79) which does exactly the same.

This pull request still improves some things in regards to `CGImage`, the most important ones are:
* Now `CGImage` uses `CGImageRelease` instead of `CFRelease` (it's a much safer to use `CGImageRelease` here, as it does not crash if you pass a `NULL` image reference to it).
* Now you can retrieve bytes from `CGImage` (uses data provider and copies the bytes from it).

Also several improvements in `CGContext`:
* Added a function `create_image()`, which allows you to create an image out of existing `CGContext`.
* Now it is possible to specify a data pointer passed as a first parameter to the `CGBitmapContextCreate()` upon `CGContext` creation -- very important thing when you deal with images as it allows you to specify your own buffer instead of allocating and copying an additional one from the context (tested it on my machine, it gave me 18.46% of performance boost).

And of course this pull requests adds `CGWindowListCreateImage()` and related things, so now window capturing is possible.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/core-graphics-rs/80)
<!-- Reviewable:end -->

application-developer-DA commented:
r? @pcwalton

jdm commented on src/geometry.rs:
```diff
@@ -70,15 +72,32 @@ impl CGRect {
+        let mut rect = CGRect::new(&CGPoint::new(0., 0.), &CGSize::new(0., 0.));
+        let result = unsafe {
+            ffi::CGRectMakeWithDictionaryRepresentation(dict.as_concrete_TypeRef(),
+                                                        &mut rect as *mut CGRect)
```
`as *mut CGRect` shouldn't be necessary.

application-developer-DA commented:
Ok, done.

I rebased and squashed the commits (if I got you right, you wanted me to rebase the commits and then to do a force push) and also I took your remark and removed an unnecessary cast in `geometry.rs`.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 81ded0d has been approved by `jdm`

<!-- @bors-servo r=jdm 81ded0d8abc5a1ca95cbfa7a7bf2c475aa1af506 -->

bors-servo commented:
:hourglass: Testing commit 81ded0d8abc5a1ca95cbfa7a7bf2c475aa1af506 with merge aa0471758a4e788f4d0b5a1d690d29020541a9bc...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/core-graphics-rs/builds/225240821)
Approved by: jdm
Pushing aa0471758a4e788f4d0b5a1d690d29020541a9bc to master...
